airflow/io/__init__.py

```diff
@@ -0,0 +1,297 @@
+        self._location = location
+
+    @abstractmethod
+    def __len__(self) -> int:
+        """Return the total length of the file, in bytes."""
+
+    @property
+    def location(self) -> str:
+        """The fully-qualified location of the input file."""
+        return self._location
+
+    @abstractmethod
+    def exists(self) -> bool:
+        """Check whether the location exists.
+
+        Raises:
+            PermissionError: If the file at self.location cannot be accessed due to a permission error.
+        """
+
+    @abstractmethod
+    def open(self, seekable: bool = True) -> InputStream:
+        """Return an object that matches the InputStream protocol.
+
+        Args:
+            seekable: If the stream should support seek, or if it is consumed sequential.
+
+        Returns:
+            InputStream: An object that matches the InputStream protocol.
+
+        Raises:
+            PermissionError: If the file at self.location cannot be accessed due to a permission error.
+            FileNotFoundError: If the file at self.location does not exist.
+        """
+
+
+class OutputFile(ABC):
+    """A base class for OutputFile implementations.
+
+    Args:
+        location (str): A URI or a path to a local file.
+
+    Attributes:
+        location (str): The URI or path to a local file for an OutputFile instance.
+        exists (bool): Whether the file exists or not.
+    """
+
+    def __init__(self, location: str):
+        self._location = location
+
+    @abstractmethod
+    def __len__(self) -> int:
+        """Return the total length of the file, in bytes."""
+
+    @property
+    def location(self) -> str:
+        """The fully-qualified location of the output file."""
+        return self._location
+
+    @abstractmethod
+    def exists(self) -> bool:
+        """Check whether the location exists.
+
+        Raises:
+            PermissionError: If the file at self.location cannot be accessed due to a permission error.
+        """
+
+    @abstractmethod
+    def to_input_file(self) -> InputFile:
+        """Return an InputFile for the location of this output file."""
+
+    @abstractmethod
+    def create(self, overwrite: bool = False) -> OutputStream:
+        """Return an object that matches the OutputStream protocol.
+
+        Args:
+            overwrite (bool): If the file already exists at `self.location`
+                and `overwrite` is False a FileExistsError should be raised.
+
+        Returns:
+            OutputStream: An object that matches the OutputStream protocol.
+
+        Raises:
+            PermissionError: If the file at self.location cannot be accessed due to a permission error.
+            FileExistsError: If the file at self.location already exists and `overwrite=False`.
+        """
+
+
+class FileIO(ABC):
+    """A base class for FileIO implementations."""
+
+    @abstractmethod
+    def new_input(self, location: str, conn_id: str | None) -> InputFile:
```

Comment:
And another, this time naming: I feel like the name `new_input()` is not correct. I would assume the file exists before, then it is not "new".
I would rather propose to name it `open()` - but this overlaps with the `FileInput` class.

Thinking out (a bit) loud... what shall be the difference between `FileIO` and `FileInput`. I assume `FileIO` is rather an entry point to "resolve" files and locations to get "File" handles or objects to operate with?
Instead of splitting "input" and "output" methods here, is it a kind of `resolve()` to a file like object (PathLike) on which you would call open (for read/write) later?

Reply:
`open` implies an action, like actually getting a file handle, which is not happening. `new_input` is a java-ism like `InputFile` and `InputStream` (remember where it originated from) and might not really be pythonic. 

Besides, I think most of that is covered by ObjectStoragePath now. I need to think about that for a bit.